src/realm/index_string.cpp

```diff
@@ -357,46 +358,44 @@ void IndexArray::index_string_all_ins(StringData value, IntegerColumn& result, C
     struct WorkItem {
         const char* header;
         size_t string_offset;
-        int permutation; // -1 == unexpanded, otherwise 0-15
+        key_type key;
     };
     using WorkList = std::vector<WorkItem>;
     WorkList work_list;
 
-    const char* top_header = get_header_from_data(m_data);
-    work_list.push_back({top_header, 0, -1});
-
     const util::Optional<std::string> upper_value = case_map(value, true);
     const util::Optional<std::string> lower_value = case_map(value, false);
 
+    auto generate_search_keys = [&work_list, &upper_value, &lower_value](const char* header,
+                                                                         size_t string_offset) -> void {
+        std::set<key_type> keys_seen;
```

Comment:
`std::set` is a tree-based data structure and very likely has to allocate a node per item it stores, resulting in the largest constant factors. `std::unordered_set` is based on a hash table and so has better time complexity on the operations that we care about while also avoiding the per-node allocation cost. `std::vector` may be faster for this use case due to being more cache-friendly, but that's speculation that would require more testing so I'd not worry about it for something that we don't intend to keep around for long.